### PR TITLE
Add `isdag` method to determine if a dask is a DAG, and add non-recursive `get`

### PR DESCRIPTION
**Pull request by eriknw**
Both of these additions need better documentation and tests, but I wanted to share.

`isdag` is a pretty standard DFS, which is stack-based (referring to data container), not recursive.

The non-recursive `get` is only used when the "get" keyword isn't given.  It can handle arbitrarily large graphs, because it is not limited by Python's recursion depth limit.  It is also faster than the recursive method.

### CODE REVIEW

File: dask/core.py [hunk 1]
```diff
@@ -41,6 +41,81 @@ def istask(x):
...
+
+        if istask(v):
+            stack = (v[0], list(v[:0:-1]), [], stack)
+            depth += 1
+            if maxdepth and depth > maxdepth:
+                if not isdag(d, list(task[1:])):
+                    raise ValueError('Cycle detected in dask!')
+                maxdepth = None
+        else:
+            results.append(v)
```
Comment (cpcloud) [inline dask/core.py]: I think you're missing a return statement.

File: dask/core.py [hunk 2]
```diff
@@ -41,6 +41,81 @@ def istask(x):
...
+                completed.add(cur)
+                nodes.pop()
+        return True
+
+
+def _get_task(d, task, maxdepth=1000):
+    # non-recursive.  DAG property is checked upon reaching maxdepth.
+    depth = 0
+    _iter = lambda *args: iter(args)
+    stack = (task[0], list(task[:0:-1]), [], None)
```
Comment (mrocklin) [inline dask/core.py]: Note that `task` may be a nested list.  We should emit results in the same configuration (e.g. possibly nested list of results).

Comment (mrocklin) [inline dask/core.py]: Ah, I see now that you do process nested lists.

File: dask/core.py (hunk 1 above)
Comment (eriknw) [inline dask/core.py]: Aha, I think the return statement got indented a level when I switched the function to work on a list of keys from a single key.

Comment (eriknw) [inline dask/core.py]: `return` for this function is on line 86.

File: dask/core.py [hunk 3]
```diff
@@ -41,6 +41,122 @@ def istask(x):
...
+
+        key = args.pop()
+        if isinstance(key, list):
+            # v = (get(d, k, concrete=False) for k in key)  # recursive
+            # Fake being lazy
+            frame = (_iter, key[::-1], [], frame)
+            depth += 1
+            if maxdepth and depth > maxdepth:
+                if not isdag(d, list(task[1:])):
+                    raise ValueError('Cycle detected in dask!')
```
Comment (eriknw) [inline dask/core.py]: Let's change this exception to either a dask-defined exception or a `RuntimeError`, which is what one would get with a recursive method.

**Pull request merged**